### ISSUE TITLE
`animateChild` does not work when used in router transitions

### ISSUE THREAD

**Issue by Ploppy3** [angular/angular#30477]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/animations

### Is this a regression?

 It never worked

### Description

`animateChild` does not work when used in router transitions as explained in the [official documentation](https://angular.io/guide/route-animations#animation-definition)

`animateChild` should start child transitions but nothing happens.

## 🔬 Minimal Reproduction

 

[stackblitz](https://stackblitz.com/edit/angular-a9hwhc?file=src%2Fapp%2Fanimations.ts)

'Content page x' should be red and slowly switch back to black when switching page.

Remove the `[@routerAnimations]="prepareRoute(outlet)"` and switch page to confirm that the animation is working but just not triggered.

## 🔥 Exception or Error
<pre><code>

</code></pre>

No error shown.

## 🌍  Your Environment

**Angular Version:**
<pre><code>

@angular@7.2.15
</code></pre>

**Anything else relevant?**

**Comment by GISerWYJ** [comment]
@Ploppy3  **This is not a bug.**
Here is a updated link. which works as you expected.
 [stackblitz](https://stackblitz.com/edit/angular-bqbf8m)
The reason why the div is not animated is when your page is in a enter state, the div element is already entered in its parent earlier ! so the '@redToBlack' bind to the div won't work as expected,because the div is not match the ':enter' state.

The right way to do is bind the 'redToBack' to the **_ page_**:
![image](https://user-images.githubusercontent.com/12789732/57904826-f4e40800-78a6-11e9-96c1-528688d167eb.png)
and modify the animation definition using query to find the div ,then animate it ！
![image](https://user-images.githubusercontent.com/12789732/57904863-19d87b00-78a7-11e9-8067-11b737ba6d80.png)

**Comment by Ploppy3** [comment]
@GISerWYJ Sorry but this is not what I want, I want the individual element in my page to render with their animation. This is a minimal reproduction of the problem. I don't want the page to be animated. Moreover if you look closely your animation has a issue where the previous page disappear instantly.

**Comment by brandon-sigao** [comment]
I'm having the exact same problem.  I've tried just about every possible way I can think of to get animations to run on the pages that are entering/leaving and it just ignores them.

**Comment by ciesielskico** [comment]
Running into the same issue

**Comment by wibimaster** [comment]
Same too :(

**Comment by michaelurban** [comment]
Can we get a status update? Is someone working on this? Are there any suggested workarounds? 

Basic functionality being broken, without comment, for 6-7 major releases is not a great position to be in.

For clarity, Angular 7 was released in October of 2018. A feature that's listed in the official docs at least four times has been broken for at least three (3) years without comment or movement. 

What's going on?

**Comment by Ploppy3** [comment]
@matsko ping

**Comment by michaelurban** [comment]
@jelbourn Could you consider bumping this to `P2` as it's such a fundamental bit of functionality.

**Comment by dario-piotrowicz** [comment]
@Ploppy3 sorry if I am getting this completely wrong, but isn't just this what you want? 
https://stackblitz.com/edit/angular-dofwv1?file=src%2Fapp%2Fanimations.ts
(just switched the query string from `':enter'` to `'@*'`)

___

Regardless I think this is probably not an issue with `animateChild` but rather with the `query` function, I can see you using `{ optional: true }` in all your queries but if you remove it from the last one (which is the problematic one) you do get an error in the console telling you that it cannot fine any such element to animate:

![Screenshot at 2021-11-15 07-51-03](https://user-images.githubusercontent.com/61631103/141743026-a8715598-0299-4726-869c-83ae97b50c2e.png)

This may be related to what @GISerWYJ mentioned, as in, the order or the elements appearing on the page.
I suspect that a parent element entering, querying for child elements entering is not working correctly
(and not sure if this is a bug or a limitation on how fundamentally elements enter the page, I'll try to look into that)

___

Also I don't think this is related to routing, as I can see the same exact behavior when I try to implement the same sort of thing using `*ngIf`s:
https://stackblitz.com/edit/angular-h7cgtl?file=src%2Fapp%2Fanimations.ts

**Comment by Ploppy3** [comment]
@dario-piotrowicz I think you're on the right way. It still seems weird to me that `@*` works but `:enter` doesn't as it exists just for this purpose.

Note: I don't work with Angular anymore so I won't be of much help anymore.

**Comment by dario-piotrowicz** [comment]
> @dario-piotrowicz I think you're on the right way. It still seems weird to me that `@*` works but `:enter` doesn't as it exists just for this purpose.
> 
> Note: I don't work with Angular anymore so I won't be of much help anymore.

@Ploppy3 Thank you very much for your input :slightly_smiling_face: 

Yes I agree with you that `@*` being ok but `:enter` not, does not feel very correct, but it sort of turns the bug from "broken functionality with no workarounds" to "mild inconvenience with an easy workaround" from my point of view :slightly_smiling_face: 

Also as I said I still want to investigate the querying issue, hopefully something interesting will turn out from that :slightly_smiling_face::+1:   

Anyway thanks a bunch, we'll take it from here then :smile:

**Comment by michaelurban** [comment]
@dario-piotrowicz Fantastic news that this may be a query issue. A mild inconvenience can be a simple hurdle to one dev and a brick wall to another, especially when the docs say it should work. 

You have no idea how grateful I am that you're working on this. 

***Mike awarded Dario:🏅(participant)***

And I'll have a freshly minted🥇1st place medal for you at this ticket's closing ceremony!

**Comment by dario-piotrowicz** [comment]
Hi, I've been spending a bunch of time on this, and I've got a bit down a rabbit hole, but I just found out something and I find it pretty interesting so I just thought of sharing it as someone could find it interesting too :smiley: 

I'm 100% that it is a query thing and it's not related to routing or anything of the sort, it is related on how angular adds elements to the dom, and based on that they can be detected/queried as entering elements for animation or not, the issue here is that the elements detected for animation are either the component's host or elements with structural directives, which our element is neither

this made me find a new workaround (if you want to use `:enter` and not `@*`, etc...), if we just put a structural directive to the child elements we want to animate, angular will just handle them differently and they will be actually detected as entering query elements

Any structural directive will do as an `*ngIf="true"` or a useless `*ngFor`:

![Screenshot at 2021-11-17 22-45-30](https://user-images.githubusercontent.com/61631103/142294601-a46aa5dc-a166-4b06-b2fe-f59740fc18b0.png)


(as you can see in this new stackblitz: https://stackblitz.com/edit/angular-lndzuf?file=src%2Fapp%2Fapp.component.html)

So yeah it's not like it's solved or anything just wanted to share this small tid-bit + it gives us another workaround to get the animation working :slightly_smiling_face:

**Comment by dario-piotrowicz** [comment]
:point_up: PR with the fix created :slightly_smiling_face: 

But:
   - I think it is a bit controversial, as I worked on it, it looked quite clear to me that the current behavior was implemented as such on purpose, so I am not 100% sure if it's beneficial to change it or it we should keep it as is but clarify the docs (I added a comment in the PR's description if you want to have a look) 
   - it's likely a bit breaky so even if it lands I am afraid it will need to do so in the next major release
   
If you want to try out the new change please pull down my branch and build the animations package yourself and link it/replace it with the node_modules one of your project :slightly_smiling_face:
(alternatively I would love to create a stackblitz with this, but am not really sure if that can be done in an easy way :thinking:, let me know if you know of any way to do so :slightly_smiling_face:)

**Comment by dario-piotrowicz** [comment]
> @dario-piotrowicz I think you're on the right way. It still seems weird to me that `@*` works but `:enter` doesn't as it exists just for this purpose.
> 
> Note: I don't work with Angular anymore so I won't be of much help anymore.

It's pretty clear to me now the `@*` (or actually any valid selection) works but `:enter` does not, as I mentioned in the PR's description that may have been intentional, since if your parent's transition is `:enter` then it is implied that all it's children are "entering" too, so if you query them in that transition you can safely assume that they are entering even if you don't query them via `:enter`

**Comment by Ploppy3** [comment]
@dario-piotrowicz I did not understand it like that at first. It makes sense now.

**Comment by petebacondarwin** [comment]
@dario-piotrowicz - thank you for looking into this and offering a fix.

I just did some reading around and playing with the examples and I have come to the conclusion that this might well be working as expected 😱.

According to this bit of the docs: https://angular.io/guide/transition-and-triggers#use-of-ngif-and-ngfor-with-enter-and-leave

>  The `:enter` transition runs when any `*ngIf` or `*ngFor` views are placed on the page, and `:leave` runs when those views are removed from the page.

So :enter and :leave refer to “views” not elements. In practice a "view" refers to the template of any structural directive (e.g. including router outlets) and not just `*ngIf` and `*ngFor`.

**Comment by petebacondarwin** [comment]
So if we think about it like this: `:enter` and `:leave` are only triggered on the root element of a view that is being added or removed, then it makes sense that querying for `:enter` on child elements of a view that is entering will not return anything.

In that case, the way I found for it to work was to just scope the query to be elements that are contained by `:enter`. In other words:

```ts
query(':enter *', [animateChild()], { optional: true }),
```

This works in the standard example we are using here: https://stackblitz.com/edit/angular-w6muoa?file=src%2Fapp%2Fanimations.ts

And if you wanted to be more precise then you could do something like:

```ts
    query(':enter @redToGreen', [animateChild()], { optional: true }),
```

See https://stackblitz.com/edit/angular-kljx26?file=src%2Fapp%2Fanimations.ts

**Comment by petebacondarwin** [comment]
I feel that we could improve the docs by defining more clearly what will receive the `:enter` and `:leave` transitions.

It seems to me that the correct definition is that these transitions will only ever appear on "views". Where a view is the instance of a template that is being added or removed by a structural directive.

Once you understand this then I think it makes sense why child elements within a template are not being marked with `:enter` when it is their view that is being added. And then providing the solution of `:enter child-selector` for marking up animation queries makes sense.

**Comment by petebacondarwin** [comment]
By the way, this would also explain why your other workaround solves the problem...

When the parent view is "entered" it will evaluate the expressions on any structural directives within its template. If any of those result in their view being displayed, those views will also receive the `:enter` transition.

**Comment by dario-piotrowicz** [comment]
> I feel that we could improve the docs by defining more clearly what will receive the `:enter` and `:leave` transitions.
> 
> It seems to me that the correct definition is that these transitions will only ever appear on "views". Where a view is the instance of a template that is being added or removed by a structural directive.
> 
> Once you understand this then I think it makes sense why child elements within a template are not being marked with `:enter` when it is their view that is being added. And then providing the solution of `:enter child-selector` for marking up animation queries makes sense.

Ok I am fine if we want to keep the behavior as is and just improve the docs accordingly

In that case I would propose to do the following:
 - close this issue as working as expected, since that seems the case and much noise is already present here
 - I'll close #44243 as that is not what we want
 - I'll amend #44216 by removing comments about things not working as expected (I would love to merge such fix asap as we are currently saying in the official docs that something is broken, which actually is not)
 - we create a new issue for the improving of the`query` documentation (I would be happy to take this on) (also, I would like to do that after #44216 as I do believe we should correct the currents docs asap)  

How would that sound? :slightly_smiling_face:

**Comment by petebacondarwin** [comment]
Yes this all sounds good. Please go ahead @dario-piotrowicz - and thanks again!!

**Comment by dario-piotrowicz** [comment]
@petebacondarwin :point_up_2: issue created, please have a look and feel free to edit its description if you feel I got something wrong or it could be improved :slightly_smiling_face:

**Comment by petebacondarwin** [comment]
Thanks @dario-piotrowicz - as discussed I will close this issue and we should continue with #44253
